lib/networkx/graph.rb

```diff
@@ -23,14 +31,37 @@ def add_edges(edges)
       end
     end
 
-    def add_node(node)
-      @nodes.add(node)
+    def add_node(node, **node_attrs)
+      if !@_nodes.key?(node)
+        @_adj[node] = {}
+        @_nodes[node] = node_attrs
+      else
+        @_nodes[node].merge!(node_attrs)
+      end
     end
 
     def add_nodes(nodes)
       case nodes
       when Set, Array
-        nodes.each { |node| add_node(node) }
+        nodes.each { |node, **node_attrs| add_node(node, node_attrs) }
+      else
+        raise ArgumentError, 'Expected argument to be an Array or Set of nodes, '\
+                             "received #{nodes.class.name} instead."
+      end
+    end
+
+    def remove_node(node)
+      @_adj[node].each_key { |k| @_adj[k].delete(node) }
+      @_adj.delete(node)
+      @_nodes.delete(node)
+    rescue KeyError, NoMethodError
+      raise KeyError, "Error in deleting node #{node} from Graph."
+    end
```

Comment:
It'd be better to have a `@nodes.key?(node)` check rather than rely on a rescue block here.